awx/api/serializers.py

```diff
@@ -347,6 +348,7 @@ def _get_related(self, obj):
 
     def _generate_named_url(self, url_path, obj, node):
         url_units = url_path.split('/')
+        reset_counters()
```

Comment:
Yea, I have the same question.

Reply:
(just answering from memory for now, can dive deeper if needed)

You know when you're doing some kind of logic on a graph, and you just have to have some state temporarily saved within the operation you're doing. Like, you know you don't need to visit something twice, so you put a `visited` counter on the objects. Well the problem is when those get left lying around because after you completed the loop or whatever they should no longer be relevant.

Due to some combination of the related links and the new middleware shim to support the old named URL, somehow we wound up calling it more than once in a request. That made the graph logic wrong.

This was definitely eyebrow-raising. There was some brittleness here that we never had to deal with before.